### DISCUSSION TITLE
Library pickle.load throws "RuntimeError: invalid value in pickle"

### DISCUSSION THREAD

**Discussion by and-tos** [rdkit/rdkit#4431] [Q&A]
Hi,

I'm trying to create a library and save it with pickle:

```
library = rdSubstructLibrary.SubstructLibrary()
for m in mols:
   m.UpdatePropertyCache()
   Chem.rdmolops.FastFindRings(m)
   m = Chem.rdmolops.AddHs(m, explicitOnly=True)
   library.AddMol(m)

with open('library.p', 'wb') as rdkit_library:
     pickle.dump(library, rdkit_library, protocol=4)
```
Trying to load it gives an error:
```
with open('library.p', 'rb') as db:
     pickle.load(db)
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
RuntimeError: invalid value in pickle
```
The code worked with other libraries, so it must be something specific to the library I'm testing. Unfortunately I cannot share the file.

Any idea what this could mean? I would really appreciate any help on the topic!

Configuration:
 - RDKit version: '2021.03.4'
 - OS: CentOS 8
 - Python version (if relevant): 3.7.10
 - Are you using conda? Yes
 - If you are using conda, which channel did you install the rdkit from? conda-forge
 - If you are not using conda: how did you install the RDKit? -

**Comment by and-tos** [comment]
The problem was caused by a molecule containing Ti, generated from mol2 format. Generating the RDKit molecule from SDF solved the problem.

**Reply by greglandrum** [reply]
That's odd. If you can share the molecule which caused the problem then we can try and fix it. Otherwise we'll just hope it doesn't come up again.

**Reply by and-tos** [reply]
Here's an example file. It's a binding site extracted from PDB 7A07. Had to change the file extension to upload it.
[7a07_test.txt](https://github.com/rdkit/rdkit/files/7167909/7a07_test.txt)

```
m = Chem.MolFromMol2File('7a07_test.txt')
with open('temp.p', 'wb+') as temp_mol:
    pickle.dump(m, temp_mol, protocol=4)    
with open(f'temp.p', 'rb') as temp_mol:
    pickle.load(temp_mol)
    
Traceback (most recent call last):
  File "/pstore/data/cadd/ccdc_roche_devel/.envs/envs/csd-python-3.0.5/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3437, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-c15177f25ccd>", line 2, in <module>
```

**Comment by drmatthewclark** [comment]
this smiles string will also cause this error in the postresql cartridge:

[O][P](N[P-2]([O])(c1ccccc1)c1ccccc1)(c1ccccc1)c1ccccc1.[Pr+6]
CC[O-].[U+5]
CC[N+](CC)(CC)CC.[Cl-].[U+5]
[Cl-].[U+5].c1ccc([As+](c2ccccc2)(c2ccccc2)c2ccccc2)cc1


it will create a molecule object, but indexing it or displaying it causes "ERROR:  molFromPickle: invalid value in pickle'

rdkit cartridge 4.1  (built from source, Nov 27 2021)

**Reply by skearnes** [reply]
@greglandrum I'm having this same issue. It doesn't seem possible to catch the error in SQL, either. I tried:

```sql
CREATE OR REPLACE FUNCTION rdkit.safe_mol_from_smiles(smiles TEXT) 
RETURNS rdkit.mol AS $$
BEGIN
    RETURN rdkit.mol_from_smiles(smiles::cstring);
EXCEPTION
WHEN OTHERS THEN
    RAISE NOTICE 'Failed to load SMILES: %s', smiles;
    RETURN NULL;
END;
$$
LANGUAGE plpgsql;
```
But I still just get the error:

```sql
orm_test=# select rdkit.mol_from_smiles('[O]P(c1ccccc1)c1ccccc1.[Pr+6]');
ERROR:  molFromPickle: invalid value in pickle
orm_test=# select rdkit.safe_mol_from_smiles('[O]P(c1ccccc1)c1ccccc1.[Pr+6]');
ERROR:  molFromPickle: invalid value in pickle
```

This is showing up in the ORM in the Open Reaction Database; current workaround is really ugly: https://github.com/open-reaction-database/ord-schema/pull/675

**Reply by skearnes** [reply]
Here's the offending SMILES: `[O-]CC.[Ti+5].[O-]CC.[O-]CC.[O-]CC.[O-]CC`

**Reply by skearnes** [reply]
(I'm on rdkit-postgresql 2021.09.2.0)

**Reply by skearnes** [reply]
(Discussion jumped to the thread below: https://github.com/rdkit/rdkit/discussions/4431#discussioncomment-4404348)

**Comment by jeremyjyang** [comment]
Here's another example, using

postgresql-14-rdkit/jammy,now 202109.2-1build1 amd64 [installed]
  Cheminformatics and machine-learning software (PostgreSQL Cartridge)

```
gsrs=> select smiles from unii_records where unii = '9S1BC7865F' ;
                smiles                
--------------------------------------
 [Cl-].[Cl-].[Cl-].[Cl-].[Cl-].[Nd+5]
(1 row)

gsrs=> select mol_from_smiles(smiles::cstring) from unii_records where unii = '9S1BC7865F' ;
ERROR:  molFromPickle: invalid value in pickle

```

**Reply by bp-kelley** [reply]
This might have been fixed here:  https://github.com/rdkit/rdkit/pull/6086

It's the only place I've seen "bad pickle value"

**Reply by skearnes** [reply]
I think it's coming from here? https://github.com/rdkit/rdkit/blob/085e20eda88cd53c0a2e1a8bb4494756b825e853/Code/GraphMol/MolPickler.cpp#L72

Also the postgresql cartridge distribution hasn't been updated for 18 months, so...

**Reply by drmatthewclark** [reply]
I recall that I worked around the issue by not having it throw that error (commenting out the test that throws the error); then it throws an error at some other point that was more catchable by the system.

**Reply by skearnes** [reply]
FYI https://github.com/conda-forge/rdkit-feedstock/issues/128

**Reply by bp-kelley** [reply]
The molecule by itself doesn't appear to have the pickle issue, if the Postgres upload is adding some props, this could trigger the issue seen in #6086

**Reply by skearnes** [reply]
FYI this does not fail on version 3.8 of the extension (the current version in my AWS Aurora instance)

**Reply by skearnes** [reply]
I can also workaround the issue locally by downgrading to `rdkit-postgresql==2020.03.3.0`
